mcswupd/main.go

```diff
@@ -0,0 +1,298 @@
+	"github.com/pkg/errors"
+)
+
+var mixWS = "/usr/share/mix"
+
+const builderConf = `[Mixer]
+LOCAL_BUNDLE_DIR = /usr/share/mix/local-bundles
+
+[Builder]
+SERVER_STATE_DIR = /usr/share/mix/update
+BUNDLE_DIR = /usr/share/mix/local-bundles
+YUM_CONF = /usr/share/mix/.yum-mix.conf
+CERT = /usr/share/mix/Swupd_Root.pem
+VERSIONS_PATH =/usr/share/mix
+LOCAL_RPM_DIR = /usr/share/mix/local-rpms
+LOCAL_REPO_DIR = /usr/share/mix/local
+
+[swupd]
+BUNDLE=os-core
+CONTENTURL=file:///usr/share/mix/update/www
+VERSIONURL=file:///usr/share/mix/update/www
+FORMAT=1
+`
+
+func getCurrentVersion() (int, error) {
+	c, err := ioutil.ReadFile("/usr/lib/os-release")
+	if err != nil {
+		return -1, err
+	}
+
+	re := regexp.MustCompile(`VERSION_ID=(\d+)\n`)
+	m := re.FindStringSubmatch(string(c))
+	if len(m) == 0 {
+		return -1, errors.New("unable to determine OS version")
+	}
+
+	v, err := strconv.Atoi(m[1])
+	if err != nil {
+		return -1, err
+	}
+
+	return v, nil
+}
+
+func setUpMixDir(bundle string, upstreamVer, mixVer int) error {
+	var err error
+	err = os.MkdirAll(filepath.Join(mixWS, "local-rpms"), 755)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(filepath.Join(mixWS, "builder.conf"),
+		[]byte(builderConf), 0644)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(filepath.Join(mixWS, "mixversion"),
+		[]byte(fmt.Sprintf("%d", mixVer)), 0644)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(filepath.Join(mixWS, "mixbundles"),
+		[]byte(fmt.Sprintf("%s\nos-core", bundle)), 0644)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filepath.Join(mixWS, "upstreamversion"), []byte(fmt.Sprintf("%d", upstreamVer)), 0644)
+}
+
+func getLastVersion() int {
+	c, err := ioutil.ReadFile(filepath.Join(mixWS, "update/image/LAST_VER"))
+	if err != nil {
+		return 0
+	}
+
+	v, err := strconv.Atoi(string(c))
+	if err != nil {
+		return 0
+	}
+
+	return v
+}
+
+// RunCommandOutput executes the command with arguments and stores its output in
+// memory. If the command succeeds returns that output, if it fails, return err that
+// contains both the out and err streams from the execution.
+func appendToFile(filename, text string) error {
+	f, err := os.OpenFile(filename, os.O_APPEND|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		_ = f.Close()
+	}()
+
+	if _, err = f.WriteString(text); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func failIfErr(err error) {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
+		os.Exit(1)
+	}
+}
+
+func buildBundles(b *builder.Builder) error {
+	if _, err := os.Stat(b.Config.Builder.Cert); os.IsNotExist(err) {
+		fmt.Println("Generating certificate for signature validation...")
+		privkey, err := helpers.CreateKeyPair()
+		if err != nil {
+			return errors.Wrap(err, "Error generating OpenSSL keypair")
+		}
+		template := helpers.CreateCertTemplate()
+		err = b.BuildBundles(template, privkey, false)
+		if err != nil {
```

Comment:
Doesn't errors.Wrap return nil if passed nil? So you can replace lines 134 to 137 with

           return errors.Wrap(b.BuildBundles(template, privkey, false), "Error building bundles")

If we just declared privkey and template as variables, we could collapse this function a bit more.

Reply:
Done, looks much better.